crates/matrix-sdk-base/src/client.rs

```diff
@@ -561,24 +563,84 @@ impl BaseClient {
         changed_devices: &api::sync::sync_events::DeviceLists,
         one_time_keys_counts: &BTreeMap<ruma::DeviceKeyAlgorithm, UInt>,
         unused_fallback_keys: Option<&[ruma::DeviceKeyAlgorithm]>,
+        changes: &mut StateChanges,
     ) -> Result<Vec<Raw<ruma::events::AnyToDeviceEvent>>> {
         if let Some(o) = self.olm_machine().await.as_ref() {
             // Let the crypto machine handle the sync response, this
             // decrypts to-device events, but leaves room events alone.
             // This makes sure that we have the decryption keys for the room
             // events at hand.
-            Ok(o.receive_sync_changes(
-                to_device_events,
-                changed_devices,
-                one_time_keys_counts,
-                unused_fallback_keys,
-            )
-            .await?)
+            let (events, room_key_updates) = o
+                .receive_sync_changes(
+                    to_device_events,
+                    changed_devices,
+                    one_time_keys_counts,
+                    unused_fallback_keys,
+                )
+                .await?;
+
+            #[cfg(feature = "experimental-sliding-sync")]
+            for room_key_update in room_key_updates {
+                if let Some(mut room) = self.get_room(&room_key_update.room_id) {
+                    self.decrypt_latest_events(&mut room, changes).await;
+                }
+            }
+
+            Ok(events)
         } else {
+            // If we have no OlmMachine, just return the events that were passed in.
+            // This should not happen unless we forget to set things up by calling
+            // set_session_meta().
             Ok(to_device_events)
         }
     }
 
+    /// Decrypt any of this room's latest_encrypted_events
+    /// that we can and if we can, change latest_event to reflect what we
+    /// found, and remove any older encrypted events from
+    /// latest_encrypted_events.
+    #[cfg(all(feature = "e2e-encryption", feature = "experimental-sliding-sync"))]
+    async fn decrypt_latest_events(&self, room: &mut Room, changes: &mut StateChanges) {
+        // Try to find a message we can decrypt and is suitable for using as the latest
+        // event. If we found one, set it as the latest and delete any older
+        // encrypted events
+        if let Some((found, found_index)) = self.decrypt_latest_suitable_event(room).await {
+            room.on_latest_event_decrypted(found, found_index);
```

Comment:
Hmm this updates the room and schedules the room info to be updated stored. But we had a `collect -> persist -> apply` cycle until now. In other words we clone the `RoomInfo`, put it into the `StateChanges` and, once persisted, we call `apply_changes()`.

I don't think that this will break anything, but it does change the way things worked.

Thinking further, since the queue of UTDs is not in `RoomInfo` it doesn't make much sense to not update the in-memory `RoomInfo` right away.

Reply:
OK, so if I understand right, this is OK as-is.

Reply:
Yes, sorry for the monologue. :sweat_smile: